datadog/resource_datadog_downtime.go

```diff
@@ -152,26 +152,77 @@ func resourceDatadogDowntime() *schema.Resource {
 	}
 }
 
-func buildDowntimeStruct(d *schema.ResourceData) *datadog.Downtime {
+// getDowntimeBoundaryTimestamp returns an int timestamp for start/end and whether or not
+// to apply this timestamp to the Downtime structure for POST/PUT request. Arguments:
+// * `d` - current `*schema.ResourceData`
+// * `dateAttr` - name of the attribute in `d` which carries and RFC3339 date string, e.g. `end_date`
+// * `tsAttr` - name of the attribute in `d` which carries integer timestamp, e.g. `end`
+// * `current` - current value (returned by API) of the boundary
+// * `updating` - whether or not this call is from Update or Create method of the downtime resource
+func getDowntimeBoundaryTimestamp(d *schema.ResourceData, dateAttr, tsAttr string, current int, updating bool) (ts int, apply bool) {
+	tsFrom := ""
+	apply = false
+
+	// first, get the int value of the boundary
+	if attr, ok := d.GetOk(dateAttr); ok {
+		if t, err := time.Parse(time.RFC3339, attr.(string)); err == nil {
+			tsFrom = dateAttr
+			ts = int(t.Unix())
+		}
+	} else if attr, ok := d.GetOk(tsAttr); ok {
+		tsFrom = tsAttr
+		ts = attr.(int)
+	}
+
+	// now decide whether we'll want to apply it or not
+	if updating {
+		// when updating, we apply when
+		// * API-returned value is different than configured value
+		// * the config value has changed
+		if current != ts || d.HasChange(tsFrom) {
+			apply = true
+		}
+	} else {
+		// when creating, we always apply
+		apply = true
+	}
+
+	return ts, apply
+}
+
+func buildDowntimeStruct(d *schema.ResourceData, client *datadog.Client, updating bool) (*datadog.Downtime, error) {
```

Comment:
Let's split this method in 2:
- One that takes care of retrieving and transforming the correct date
- One that deals with understanding if we should update or not. Let's document this one as well a little bit more to better explain why we need this.

Reply:
Done, please re-review. Thanks for the feedback!